working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+
+If no hashCode getter is defined in or inherited by the struct (except from
+Object), then a hashCode getter will be defined which hashes the runtime type
+(again, actual type) together with the hashes of each of the fields.
+
+#### toString
+
+If no toString method is defined in or inherited by the struct (except from
+Object), then a toString method will be defined which returns "struct".
+
+#### debugToString
+
+If no debugToString method is defined in or inherited by the struct, then a
+debugToString method will be defined which returns a formatted description of
+the receiver, of the form `<type>(<f0>, ..., <f1>)` where `<type>` is the
+runtime type of the receiver, and the `fi` are the result of calling
+"debugToString" on the `i`th field if that field is (dynamically) a struct, and
+otherwise the result of calling "toString".
+
+**COMMENTARY(leafp):** *This needs a bit of work.  If we keep the dynamic check
+  for "struct-ness", then we need do deal with the case that the user defines a
+  debutToString thing with the wrong type*.
+
+**COMMENTARY(leafp):** *If we keep this, we may wish to specify that compilers
+  may choose to make it a link time error to have an invocation of debugToString
+  in the program outside of asserts, and other debutToString methods*.
+
+
+#### copyWith
+
+If no copyWith method is defined in or inherited by the struct, then a copyWith
+method will be defined with named parameters for every entry in the primary
+constructor.  For every entry, if the field name is not private, then the
+parameter name is the field name.  If the field name is private, then the
+parameter name is the field name with all leading `_`s removed.
+
+The body of the method calls the generated primary constructor.  For every
+argument which is explicitly passed to the copyWith method invocation, the
+corresponding parameter is passed on as the argument to the corresponding
+parameter of the generated primary constructor.  For every argument which is not
+passed to the copyWith method invocation, the value of the corresponding field
+from the current instance is passed on as the argument to the corresponding
+parameter of the generated primary constructor.
+
+**COMMENTARY(leafp):** *As with constructors, this method cannot be generated as
+  a strict de-sugaring, since the ability to detect whether or not an argument
+  has been passed is not available in Dart.  Implementations already support
+  this for default values, so it is likely not problematic to implement*.
+
+
+#### Additional generated members
+
+We may wish to also define additional generated members.  For example, we may
+wish to have `parse` and `unParse` methods.  TODO(leafp): consider fleshing
+something out here.
+
+
+#### Restrictions
+
+It is an error to override the "runtimeType" method of a struct.
+
+It is an error to override the "noSuchMethod" method of a struct.
+
+### Const structs
+
+TODO(leafp): This should work, write out the details.
+
+## Extension Structs
```

Comment:
One important constraint we'd like it to have explicitly specified is that we need these extensions to be "sticky", that is, they should be available as part of the type of the expression, and not just based on scoping rules.

Because of this, I'm also not sure that the word `extension` maps well with the existing use of that keyword with classes. 

An idea here could be to use the word `static` instead:

```dart
static struct F(x) {
  int get foo => x.bar;
}
```

`static` here means that the dispatch is static, like extension methods; but also that it is a static (compile-time only) concept and the struct and type name is erased at runtime.

Reply:
Yes, this is the semantics (I'm taking this as a given for this discussion).  I also like static quite a bit for the modifier.

Reply:
Filed #2363 to discuss syntax.